### PR TITLE
getHostType: consider APP_DOC_INTERNAL_URL as native

### PR DESCRIPTION
**Pull request by fflorent**
# Context

While trying to scale, we faced some issue where we have a different internal and public URL for doc workers, and having configured the org to be specified in the path (`GRIST_ORG_IN_PATH=true`).
We saw that we can set `APP_DOC_INTERNAL_URL` in the doc worker so the home worker knows how to contact to the doc workers through a private network.

However, without this patch, when the home workers attempts to contact the doc worker, the latter responds a 404 error and with a `Domain not recognized: ${hostname}` error message. That's because of the code below which rely on the result of the `getHostType()` function. I am not sure of what how the machinery works here when the `getHostType()` equals to custom though.

https://github.com/gristlabs/grist-core/blob/9557f8c4e4c61b434ee508d4fedcec063209d900/app/server/lib/extractOrg.ts#L94-L121

# Proposed solution

If we tell `getHostType()` to return `"native"` when the host corresponds to the value of `APP_DOC_INTERNAL_URL`, the rest works.
As I said, I am not very certain of the implications though. Feedback welcome :pray:

### CODE REVIEW

File: test/common/gristUrls.ts [hunk 1]
```diff
@@ -64,4 +64,44 @@ describe('gristUrls', function() {
...
     });
   });
+
+  describe('getHostType', function() {
+    const defaultOptions = {
+      baseDomain: 'getgrist.com',
+      pluginUrl: 'https://plugin.getgrist.com/path',
+    };
+    afterEach(function () {
+      delete process.env.APP_DOC_INTERNAL_URL;
```
Comment (berhalak) [inline test/common/gristUrls.ts:74]: Take a look in `DocApi2.ts` tests, we have an utility class to work with env variables `new testUtils.EnvironmentSnapshot();`

File: test/common/gristUrls.ts [hunk 2]
```diff
@@ -64,4 +64,44 @@ describe('gristUrls', function() {
       assert.deepEqual(parseFirstUrlPart('o', ''), {path: ''});
     });
   });
+
+  describe('getHostType', function() {
+    const defaultOptions = {
+      baseDomain: 'getgrist.com',
+      pluginUrl: 'https://plugin.getgrist.com/path',
```
Comment (berhalak) [inline test/common/gristUrls.ts:71]: Why the extra path at the end? I think Grist assumes (silently?) that this is an absolute URL (protocol + host) only. At least I never tested or see it being used with a path.

**Review by berhalak — COMMENTED** [review]
Looks ok to me, but I'll add @paulfitz  as a reviewer, just to be sure. I'm not entirely confidant that I understand that part of the system.

File: test/common/gristUrls.ts (hunk 2 above)
Comment (fflorent) [reply test/common/gristUrls.ts:71]: Thanks for your feedback!

I am not sure of the purpose of this variable (nor of `APP_UNTRUSTED_URL`), so I inferred that test from what I saw in the code. 

From what I see in your question, I will remove the path and only keep the protocol + host.

File: test/common/gristUrls.ts (hunk 1 above)
Comment (fflorent) [reply test/common/gristUrls.ts]: Thanks! I took into account all your remarks in this commit: 3aa126133fea7dc372c229c0f3b06f351cb6cd40

File: test/common/gristUrls.ts [hunk 3]
```diff
@@ -68,29 +69,36 @@ describe('gristUrls', function() {
   describe('getHostType', function() {
     const defaultOptions = {
       baseDomain: 'getgrist.com',
-      pluginUrl: 'https://plugin.getgrist.com/path',
+      pluginUrl: 'https://plugin.getgrist.com',
     };
+
+    const oldEnv = new testUtils.EnvironmentSnapshot();
```
Comment (berhalak) [inline test/common/gristUrls.ts:75]: This must be inside a `before` event.

Comment (fflorent) [reply test/common/gristUrls.ts]: Done!

I assume that in DocApi2.ts, this should the same.

File: app/common/gristUrls.ts [hunk 4]
```diff
@@ -163,6 +163,13 @@ export interface OrgUrlInfo {
   orgInPath?: string;     // If /o/{orgInPath} should be used to access the requested org.
 }
 
+function isDocInternalUrl(host: string) {
+  if (!process.env.APP_DOC_INTERNAL_URL) { return false; }
+  const internalUrl = new URL('/', process.env.APP_DOC_INTERNAL_URL);
+  const internalHostnameAndMaybePort = internalUrl.hostname + (internalUrl.port ? `:${internalUrl.port}` : '');
```
Comment (paulfitz) [inline app/common/gristUrls.ts:169]: What is the difference between this and `internalUrl.host` (not `hostname`)?

**Review by paulfitz — COMMENTED** [review]
This narrow change seems reasonable to me.

File: README.md [hunk 5]
```diff
@@ -250,6 +250,7 @@ Variable | Purpose
 -------- | -------
 ALLOWED_WEBHOOK_DOMAINS | comma-separated list of permitted domains to use in webhooks (e.g. webhook.site,zapier.com). You can set this to `*` to allow all domains, but if doing so, we recommend using a carefully locked-down proxy (see `GRIST_HTTPS_PROXY`) if you do not entirely trust users. Otherwise services on your internal network may become vulnerable to manipulation.
 APP_DOC_URL | doc worker url, set when starting an individual doc worker (other servers will find doc worker urls via redis)
+APP_DOC_INTERNAL_URL | like `APP_DOC_URL` but used by the home server to join the server using an internal domain name resolution (like in a docker environment). Defaults to APP_DOC_URL
```
[the remark this answers is not in the document]
Comment (fflorent) [reply README.md]: Thanks!

File: app/common/gristUrls.ts (hunk 4 above)
Comment (fflorent) [reply app/common/gristUrls.ts]: You're right, it works as well (and that's much simpler! :) ). Thanks for the hint, that's done

**Review by berhalak — APPROVED** [review]
Thanks @fflorent

**Pull request merged**